src/ray/object_manager/object_buffer_pool.cc

```diff
@@ -75,13 +75,13 @@ std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status> ObjectBufferPool::Ge
                                                  object_buffer.metadata->Size()));
     auto *data = object_buffer.data->Data();
     uint64_t num_chunks = GetNumChunks(data_size);
-    get_buffer_state_.emplace(
-        std::piecewise_construct, std::forward_as_tuple(object_id),
-        std::forward_as_tuple(BuildChunks(object_id, data, data_size)));
+    get_buffer_state_.emplace(std::piecewise_construct, std::forward_as_tuple(object_id),
+                              std::forward_as_tuple(BuildChunks(
+                                  object_id, data, data_size, object_buffer.data)));
     RAY_CHECK(get_buffer_state_[object_id].chunk_info.size() == num_chunks);
   }
   get_buffer_state_[object_id].references++;
-  return std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status>(
+  return std::pair<const ObjectBufferPool::ChunkInfo, ray::Status>(
       get_buffer_state_[object_id].chunk_info[chunk_index], ray::Status::OK());
 }
 
```

Comment:
for my understanding: it seems to me it's not the `Buffer` who keeps the underlining object's lifetime but rather the `ReleaseGetChunk` will calls back to plasma to release the memory for this object if reference counting drops to 0.
So if the understanding is correct it means 

- if the `Buffer` is still in scope, if we have a bug prematurely release the reference we could still encounter this memory corruption issue.
- if the `Buffer` is not in scope, as long as this reference counting is correct the memory access should be safe.

Reply:
> not the Buffer who keeps the underlining object's lifetime

I don't think that's right, here in client.cc you can see that PlasmaBuffer's destructor calls client->Release(): https://github.com/ray-project/ray/blob/master/src/ray/object_manager/plasma/client.cc#L211

As to why there's an extra Release() call in ReleaseGetChunk, I believe that's because there's an extra initial ref when creating an object, before the buffer exists.

Reply:
I see, makes sense.